runtime/src/main/cpp/Memory.h

```diff
@@ -315,16 +315,27 @@ ALWAYS_INLINE bool hasPointerBits(T* ptr, unsigned bits) {
 struct MetaObjHeader {
   // Pointer to the type info. Must be first, to match ArrayHeader and ObjHeader layout.
   const TypeInfo* typeInfo_;
-  // Strong reference to the counter object.
-  ObjHeader* counter_;
   // Container pointer.
   ContainerHeader* container_;
+
 #ifdef KONAN_OBJC_INTEROP
   void* associatedObject_;
 #endif
 
   // Flags for the object state.
   int32_t flags_;
+
+  // TODO: maybe make it a union for the orthogonal features.
+  struct {
+    // Strong reference to the counter object.
+    ObjHeader* counter_;
+  } WeakReference;
+  struct {
+    // Leak detector's previous list element.
+    ObjHeader* previous_;
+    // Leak detector's next list element.
+    ObjHeader* next_;
+  } LeakDetector;
```

Comment:
This makes Objective-C interoperability consume even more memory.

Reply:
Indeed.

Reply:
We may want to completely separate debug and release runtimes bitcodes, and make those debug-only.

Reply:
Or meta object allocator could reserve extra room for these fields in debug mode.